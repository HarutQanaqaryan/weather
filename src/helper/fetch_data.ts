const API_KEY = `6a7c24bc3b75b3a53d16bcec8f105fec`;


const getData = (cityName: string, count: number): Promise<any> => {
   return fetch(
        `https://api.openweathermap.org/data/2.5/forecast?q=${cityName}&units=metric
              &cnt=${count}&appid=${API_KEY}`
      )
}

export default getData;