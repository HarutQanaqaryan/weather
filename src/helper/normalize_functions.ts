import convertDate from "./convert_date";

export const normalizeDate = (seconds: number): string => {
  const normalizeTime = (time: number) => (time <= 9 ? "0" + time : time);
  let date = new Date(seconds);
  return `${date.getDate()} ${convertDate(
    date.getMonth()
  )} ${date.getFullYear()} ${date.getHours()}:${normalizeTime(
    date.getUTCMinutes()
  )}`;
};

export const normalizeTemp = (temp: number): number => {
  return Math.round(temp - 273.15);
};
