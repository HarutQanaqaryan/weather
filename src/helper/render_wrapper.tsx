import RenderWeatherCard from "./render_weather_card";

interface RenderWrapperProps {
  loading: boolean;
  cityName: string;
  data: any[]
}

const RenderWrapper = ({ loading, cityName, data }: RenderWrapperProps) => {
  return (
    <div className="wrapper_weather">
      <h2 className="city_name_header">{cityName}</h2>
      {loading && <div className="loading"></div>}
      <div className="data_block">{<RenderWeatherCard data={data} />}</div>
    </div>
  );
};

export default RenderWrapper;
