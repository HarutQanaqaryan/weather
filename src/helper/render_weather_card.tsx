import { useState, useEffect, FC, MouseEvent } from "react";
import { normalizeDate, normalizeTemp } from "./normalize_functions";

interface RenderWeatherCardProps {
  data: any[]
}

const RenderWeatherCard: FC<RenderWeatherCardProps> = ({ data }) => {
  const [isOpen, setIsOpen] = useState<any>();

  useEffect(() => {
    setIsOpen(data.map((_: any, idx: number) => ({ id: idx, open: false })));
  }, [data]);


  const openMoreDetails = (event: MouseEvent<HTMLButtonElement>) => {
    setIsOpen(
      isOpen.map((el: any) =>
        el.id === parseInt(event.currentTarget.id) ? { ...el, open: !el.open } : el
      )
    );
  };

  return (
    <div className="data_block">
      {data.map((el: any, idx: any) => (
      <div key={idx} className="weather_card">
          <h4 className="date">{normalizeDate(el.dt * 1000)}</h4>
          <img
            src={`https://openweathermap.org/img/wn/${el.weather[0].icon}@2x.png`}
            alt="Weather Icon"
            className="weather_icon"
          />
          <h1 className="main_temp">{normalizeTemp(el.main.temp)}째c</h1>
          <h5 className="weather_desctription">{el.weather[0].description}</h5>
          <h5>
            <span className="desctription_header">Feels Like </span>
            {normalizeTemp(el.main.feels_like)}째C
          </h5>
        <button className="more_details_btn" id={idx} onClick={openMoreDetails}>
          More Details
        </button>
        {isOpen.some((el: any) => el.open && el.id === idx) && (
          <div className="weather_desctriptions">
            <b className="temp_header">Min Temp </b>
            <span className="temp">{normalizeTemp(el.main.temp_min)}째c</span>
            <br />
            <b className="temp_header">Max Temp </b>
            <span className="temp">{normalizeTemp(el.main.temp_max)}째c</span>
            <b>
              <span className="desctription_header">Humidity </span>
              {el.main.humidity}%
            </b>
            <b>
              <span className="desctription_header">Pressure </span>
              {el.main.pressure}hPa
            </b>
          </div>
        )}
      </div>
      ))}
    </div>
  );
};

export default RenderWeatherCard;
