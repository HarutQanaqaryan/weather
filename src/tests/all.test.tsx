import renderer from "react-test-renderer";
import convertDate from "../helper/convert_date";
import getData from "../helper/fetch_data";
import { normalizeDate, normalizeTemp } from "../helper/normalize_functions";
import RenderWeatherCard from "../helper/render_weather_card";
import GetWeather from "../components/home_page_weather";
import RenderWrapper from "../helper/render_wrapper";
import SearchByCityName from "../components/search_by_city_name";

export {};

test("Convertation Date", () => {
  expect(convertDate(1)).toBe("Feb");
});

test("Get Data", () => {
  return getData("Minsk", 17)
    .then((r) => r.json())
    .then((data) => {
      data.list.map((el: {}) => expect(el).toBe(el));
    });
});

test("Normalize Date", () => {
  expect(normalizeDate(1637085600)).toBe("20 Jan 1970 1:44");
});

test("Normalize Temp", () => {
  expect(normalizeTemp(272.81)).toBe(-0);
});

const fetchDataForTesting = (fetch_data: any) => {
  let DATA: any[] = [];
  fetch_data
    .then((r: any) => r.json())
    .then((data: any) => data.list.map((el: {}) => DATA.push(el)));
  return DATA;
};

test("Render Weather Card", () => {
  const tree = renderer
    .create(
      <RenderWeatherCard data={fetchDataForTesting(getData("Minsk", 17))} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Render Wrapper", () => {
  const tree = renderer
    .create(
      <RenderWrapper
        loading={false}
        cityName={"Minsk"}
        data={fetchDataForTesting(getData("Minsk", 17))}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Get Weather", () => {
  const tree = renderer.create(<GetWeather />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Search By CityName", () => {
  const tree = renderer.create(<SearchByCityName />).toJSON();
  expect(tree).toMatchSnapshot();
});