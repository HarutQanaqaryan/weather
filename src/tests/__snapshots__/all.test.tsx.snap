// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Weather 1`] = `
<div>
  <button
    className="city_name_btn"
    onClick={[Function]}
  >
    Minsk
  </button>
  <button
    className="city_name_btn"
    onClick={[Function]}
  >
    Bratislava
  </button>
  <button
    className="city_name_btn"
    onClick={[Function]}
  >
    Moscow
  </button>
  <div
    className="wrapper_weather"
  >
    <h2
      className="city_name_header"
    >
      Minsk
    </h2>
    <div
      className="data_block"
    >
      <div
        className="data_block"
      />
    </div>
  </div>
</div>
`;

exports[`Render Weather Card 1`] = `
<div
  className="data_block"
/>
`;

exports[`Render Wrapper 1`] = `
<div
  className="wrapper_weather"
>
  <h2
    className="city_name_header"
  >
    Minsk
  </h2>
  <div
    className="data_block"
  >
    <div
      className="data_block"
    />
  </div>
</div>
`;

exports[`Search By CityName 1`] = `
Array [
  <input
    className="search"
    onChange={[Function]}
    placeholder="Type city name..."
    value="Minsk"
  />,
  <div
    className="wrapper_weather"
  >
    <h2
      className="city_name_header"
    >
      Minsk
    </h2>
    <div
      className="loading"
    />
    <div
      className="data_block"
    >
      <div
        className="data_block"
      />
    </div>
  </div>,
]
`;
